tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/FirebaseSubscription.scala

```diff
@@ -166,10 +172,96 @@ object FirebaseSubscriptionCreation {
+    case _ => None
+  }).toMap
+}
+
+object FirebaseSubscriptionPatchObject {
+  private val updateProperties: Set[String] = Set("types", "expires")
+}
+
+case class FirebaseSubscriptionPatchObject(value: Map[String, JsValue]) {
+
+  def validate(typeStateFactory: TypeStateFactory): Either[FirebasePatchUpdateValidationException, ValidatedFirebaseSubscriptionPatchObject] =
+    for {
+      validatedProperties <- validateProperties
+      types <- validateTypes(typeStateFactory)
+    } yield {
+      ValidatedFirebaseSubscriptionPatchObject.from(types)
+    }
+
+  def validateProperties: Either[FirebasePatchUpdateValidationException, FirebaseSubscriptionPatchObject] =
+    value.find(pair => !updateProperties.contains(pair._1))
+      .map(e => scala.Left(FirebasePatchUpdateValidationException("Some unknown properties were specified", Some(e._1))))
+      .getOrElse(scala.Right(this))
+
+  def validateTypes(typeStateFactory: TypeStateFactory): Either[FirebasePatchUpdateValidationException, List[TypeName]] =
+    value.get("types") match {
+      case Some(jsValue) => jsValue match {
+        case JsArray(arrayValue) => arrayValue.toList
+          .map(typeValue => parseType(typeValue, typeStateFactory))
+          .traverse(x => Validated.fromEither(x.left.map(List(_)))).toEither
+          .left.map(_.head)
+          .flatMap(validateEmptyArrayTypes)
+        case _ => Left(FirebasePatchUpdateValidationException("Expecting an array of JSON strings as an argument", Some("types")))
+      }
+      case None => scala.Right(List())
+    }
+
+  def validateEmptyArrayTypes(list: List[TypeName]): Either[FirebasePatchUpdateValidationException, List[TypeName]] =
+    if (list.isEmpty) {
+      scala.Left(FirebasePatchUpdateValidationException("Must not empty", Some("types")))
+    } else {
+      scala.Right(list)
+    }
+
+  private def parseType(jsValue: JsValue, typeStateFactory: TypeStateFactory): Either[FirebasePatchUpdateValidationException, TypeName] = jsValue match {
+    case JsString(aString) => typeStateFactory.parse(aString).left.map(e => FirebasePatchUpdateValidationException(e.getMessage, Some("types")))
+    case _ => Left(FirebasePatchUpdateValidationException("Expecting an array of JSON strings as an argument", Some("types")))
+  }
+}
+
+object ValidatedFirebaseSubscriptionPatchObject {
+  def from(typeNames: List[TypeName]): ValidatedFirebaseSubscriptionPatchObject =
+    if (typeNames.isEmpty) {
+      ValidatedFirebaseSubscriptionPatchObject()
+    } else {
+      ValidatedFirebaseSubscriptionPatchObject(Some(typeNames.toSet))
+    }
+}
+
+case class ValidatedFirebaseSubscriptionPatchObject(typeUpdate: Option[Set[TypeName]] = None)
+
+case class FirebaseSubscriptionSetRequest(create: Option[Map[FirebaseSubscriptionCreationId, JsObject]],
+                                          update: Option[Map[UnparsedFirebaseSubscriptionId, FirebaseSubscriptionPatchObject]]) extends WithoutAccountId
```

Comment:
Why did you move down the setRequest object?

Reply:
just for move all update model to bottom of class (easier for follow)

Reply:
Hmmmm... if you say so :)